src/validator/validator.ts

```diff
@@ -0,0 +1,136 @@
+  private validatorIndex: ValidatorIndex;
+  private blockService: BlockProcessingService;
+  private genesisInfo: GenesisInfo;
+  public isActive: boolean;
+
+  /**
+   * @param {ValidatorCtx} ctx
+   */
+  public constructor(ctx: ValidatorCtx) {
+    this.ctx = ctx;
+    this.logger = logger;
+    this.isActive = false;
+  }
+
+  /**
+   * Creates a new block proccessing service and starts it.
+   */
+  private async start(): Promise<void> {
+    await this.setup();
+    this.run();
+  }
+
+  /**
+   * Main method that starts a client.
+   * @returns {Promise<void>}
+   */
+  public async setup(): Promise<void> {
+    this.logger.info("Setting up validator client...");
+
+    await this.setupRPC();
+
+    // Wait for the ChainStart log and grab validator index
+    this.isActive = await this.isChainLive();
+    this.validatorIndex = await this.getValidatorIndex();
+
+    await this.setupServices();
+  }
+
+  /**
+   * Establishes a connection to a specified beacon chain url.
+   */
+  private setupRPC(): void {
+    this.logger.info("Setting up RPC connection...");
+    this.provider = new RPCProvider(this.ctx.rpcUrl);
+    this.logger.info(`RPC connection successfully established ${this.ctx.rpcUrl}!`);
+  }
+
+  /**
+   * Recursively checks for the chain start log event from the ETH1.x deposit contract
+   * @returns {Promise<boolean>}
+   */
+  private async isChainLive(): Promise<boolean> {
+    this.logger.info("Checking if chain has started...");
+    if (await this.provider.hasChainStarted()) {
+      this.genesisInfo = await this.provider.getGenisisInfo();
+      this.logger.info("Chain start has occured!");
+      return true;
+    }
+    setTimeout(this.isChainLive, 1000);
+  }
+
+  /**
+   * Checks to see if the validator has been processed on the beacon chain.
+   * @returns {Promise<ValidatorIndex>}
+   */
+  private async getValidatorIndex(): Promise<ValidatorIndex> {
+    this.logger.info("Checking if validator has been processed...");
+    const index = await this.provider.getValidatorIndex(this.ctx.publicKey);
+    if (index) {
+      this.logger.info("Validator has been processed!");
+      return index;
+    }
+    setTimeout(this.getValidatorIndex, 1000);
+  }
+
+  /**
+   * Setups the necessary services.
+   * @returns {Promise<void>}
+   */
+  private async setupServices(): Promise<void> {
+    this.blockService = new BlockProcessingService(this.validatorIndex, this.provider, this.ctx.privateKey);
+    // TODO setup attestation service
+  }
+
+  private async checkAssignment(): Promise<void> {
+    // If epoch boundary then look up for new assignment
+    if ((Date.now() - this.genesisInfo.startTime) % SLOTS_PER_EPOCH === 0) {
+      const epoch = await this.provider.getCurrentEpoch();
+      const assignment: CommitteeAssignment = await this.provider.getCommitteeAssignment(epoch, this.validatorIndex);
+      const isProposer: boolean = this.provider.isProposerAtSlot(assignment.slot, this.validatorIndex);
```

Comment:
it seems like this is relying on the beacon chain to calculate if the validator is a proposer at a given slot.
I think we should have the validator calculate if they're a proposer for a given slot.

Reply:
Right, but that would mean the validator needs to maintain a pruned copy of the state in order to make that calculation. I can check in on how other clients do these opperations, but I'm curious as to what you think.

IMO thats definitely the next iteration anyway - Need to add a DB to the validator to catch slashing conditions and also that would help facilitate this.